File: 0872-leaf-similar-trees/0872-leaf-similar-trees.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public void takeLeaves(TreeNode current,List<int> leafList)
    {
        if(current==null){return;}
        else if(current.left==null && current.right==null)
        {
            leafList.Add(current.val);
        }
        else
        {
            takeLeaves(current.left,leafList);
            takeLeaves(current.right,leafList);
        }
    }
    public bool LeafSimilar(TreeNode root1, TreeNode root2) {
        List<int>firstLeafList=new List<int>();
        List<int>secondLeafList=new List<int>();
        takeLeaves(root1,firstLeafList);
        takeLeaves(root2,secondLeafList);
        bool areEqual = firstLeafList.SequenceEqual(secondLeafList);
        return areEqual;
    }
}

File: 0872-leaf-similar-trees/0872-leaf-similar-trees.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private List<Integer> firstLeaves = new ArrayList<>();
    private List<Integer> secondLeaves = new ArrayList<>();
    
    private void helper(TreeNode current, boolean isFirst){
        if(current == null) return;
        if(current.left == null && current.right == null) {
            if(isFirst) {
               firstLeaves.add(current.val); 
            } else {
               secondLeaves.add(current.val);
            }
            return;
        }
        
        helper(current.left,isFirst);
        helper(current.right,isFirst);
    }
    
    public boolean leafSimilar(TreeNode root1, TreeNode root2) {
        helper(root1,true);
        helper(root2,false);
        
        if(firstLeaves.size() != secondLeaves.size()) return false;
        
        for(int i=0;i<firstLeaves.size();i++){
            if(!firstLeaves.get(i).equals(secondLeaves.get(i))) return false;
        }
        
        return true;
    }
}